src/main/scala/mesosphere/marathon/api/akkahttp/v2/GroupsController.scala

```diff
@@ -39,50 +63,101 @@ class GroupsController(electionService: ElectionService, infoService: GroupInfoS
+    }
+  }
+
+  /**
+    * Until there is async version of updateRoot we have to block here
+    */
+  protected def result[T](fn: Awaitable[T]): T = Await.result(fn, config.zkTimeoutDuration)
+
+  @SuppressWarnings(Array("all")) // async/await
+  private def updateOrCreate(id: PathId, update: raml.GroupUpdate, force: Boolean)(implicit identity: Identity): Future[DeploymentResult] = async {
+    val version = Timestamp.now()
+
+    val effectivePath = update.id.map(PathId(_).canonicalPath(id)).getOrElse(id)
+    val deploymentPlan = await(groupManager.updateRoot(
+      id.parent, group => result(groupApiService.updateGroup(group, effectivePath, update, version)), version, force))
+    DeploymentResult(deploymentPlan.id, deploymentPlan.version.toOffsetDateTime)
+  }
 
   def updateGroup(groupId: PathId): Route = ???
 
   def deleteGroup(groupId: PathId): Route = ???
 
-  def listVersions(groupId: PathId): Route = ???
+  def listVersions(groupId: PathId)(implicit identity: Identity): Route = {
+    groupManager.group(groupId) match {
+      case Some(group) =>
+        authorized(ViewGroup, group).apply {
+          val versionsFuture = groupManager.versions(groupId).runWith(Sink.seq)
+          onSuccess(versionsFuture) { versions =>
+            complete(versions)
+          }
+        }
+      case None => reject(EntityNotFound.noGroup(groupId))
+    }
+  }
 
-  def versionDetail(groupId: PathId, version: String): Route = ???
+  def versionDetail(groupId: PathId, version: Timestamp)(implicit identity: Identity): Route = extractEmbeds {
+    case (appEmbed, groupEmbed) =>
+      onSuccess(infoService.selectGroupVersion(groupId, version, authorizationSelectors, groupEmbed)) {
+        case Some(info) => complete(Json.toJson(info))
+        case None => reject(EntityNotFound.noGroup(groupId, Some(version)))
+      }
+  }
 
   // format: OFF
   val route: Route = {
     asLeader(electionService) {
       authenticated.apply { implicit identity =>
-        path(GroupPathIdLike) { groupId =>
-          pathEndOrSingleSlash {
-            get {
-              groupDetail(groupId)
-            } ~
-            post {
-              createGroup(groupId)
-            } ~
-            put {
-              updateGroup(groupId)
-            } ~
-            delete {
-              deleteGroup(groupId)
+        path(GroupPathIdLike) { maybeGroupId =>
+          withValidatedPathId(maybeGroupId.toString) { groupId =>
+            pathEndOrSingleSlash {
+              get {
+                groupDetail(groupId)
+              } ~
+              post {
+                createGroup(groupId)
+              } ~
+              put {
+                updateGroup(groupId)
+              } ~
+              delete {
+                deleteGroup(groupId)
+              }
             }
           }
         } ~
-        path(GroupPathIdLike ~ "apps" ~ Slash.? ~ PathEnd) { groupId =>
+        path(GroupPathIdLike ~ Slash.? ~ "apps" ~ Slash.? ~ PathEnd) { groupId =>
```

Comment:
I think there is a `PathEndOrSlash`, isn't there?

Reply:
but this is not path end, right... ?

Reply:
What do you think about moving this into a separate `PathMatcher`?

Reply:
@meln1k not sure I understand... Slash.? is already a path matcher :)

Reply:
I meant the last part `~ Slash.? ~ PathEnd`.